## 6.1 Sourcing Open Data
## - Hotel Booking Analysis

## 1. Importing Libraries and Dataset

## 2. Cleaning Data

## 3. Renaming columns

## 4. Data Understanding

# 1.Importing Libraries and Dataset

In [2]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [3]:
# Importing orders data - using os library
path = r'C:\Users\35196\Documents\Alexandra\Data Analytics Program\02 Immersion\6. Advanced Analytics and Dashboard Design\Hotel Bookings Analysis'

df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'hotel_bookings.csv'), index_col = False)

# 2. Cleaning Data

In [4]:
# Exploring Data 
df.head()

index         hotel  is_canceled  lead_time  arrival_date_year  \
0      0  Resort Hotel            0        342               2015   
1      1  Resort Hotel            0        737               2015   
2      2  Resort Hotel            0          7               2015   
3      3  Resort Hotel            0         13               2015   
4      4  Resort Hotel            0         14               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  deposit_type  agent  \
0                        0                     0  ...    No Deposit    NaN   
1                        0                     0  ...    No Deposit    NaN   
2                        0                     1  ...    No Deposit    NaN   
3                        0                     1  ...    No Deposit  304.0   
4                        0                     2  ...    No Deposit  240.0   

   company days_in_waiting_list customer_type   adr  \
0      NaN                    0     Transient   0.0   
1      NaN                    0     Transient   0.0   
2      NaN                    0     Transient  75.0   
3      NaN                    0     Transient  75.0   
4      NaN                    0     Transient  98.0   

  required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                           0                          0           Check-Out   
1                           0                          0           Check-Out   
2                           0                          0           Check-Out   
3                           0                          0           Check-Out   
4                           0                          1           Check-Out   

   reservation_status_date  
0                 01-07-15  
1                 01-07-15  
2                 02-07-15  
3                 02-07-15  
4                 03-07-15  

[5 rows x 33 columns]

In [5]:
df.shape

(119390, 33)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 33 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   index                           119390 non-null  int64  
 1   hotel                           119390 non-null  object 
 2   is_canceled                     119390 non-null  int64  
 3   lead_time                       119390 non-null  int64  
 4   arrival_date_year               119390 non-null  int64  
 5   arrival_date_month              119390 non-null  object 
 6   arrival_date_week_number        119390 non-null  int64  
 7   arrival_date_day_of_month       119390 non-null  int64  
 8   stays_in_weekend_nights         119390 non-null  int64  
 9   stays_in_week_nights            119390 non-null  int64  
 10  adults                          119390 non-null  int64  
 11  children                        119386 non-null  float64
 12  babies          

In [7]:
# Checking missing values
df.isnull().sum()

index                                  0
hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent           

In [8]:
# I can conclude the dataset is clean from missing values.
# Agent and company have significant missing values, however this variables are not relevant for the analysis.
# Agent is 'the ID of the travel agency that made the booking'
# company is 'The ID of the company that made the booking'

# Let's delete these variables

In [9]:
df.drop(['company','agent'], axis =1, inplace = True)

In [10]:
df.isnull().sum()

index                               0
hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_year                   0
arrival_date_month                  0
arrival_date_week_number            0
arrival_date_day_of_month           0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            4
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
assigned_room_type                  0
booking_changes                     0
deposit_type                        0
days_in_waiting_list                0
customer_type                       0
adr         

 Now only 4 missing values for children variable
 and 488 missing values for country, that represents less than 0.5% of the observations -> Let's keep it as it is

In [11]:
# Checking duplicates
df_dups=df[df.duplicated()]

In [12]:
df_dups

Empty DataFrame
Columns: [index, hotel, is_canceled, lead_time, arrival_date_year, arrival_date_month, arrival_date_week_number, arrival_date_day_of_month, stays_in_weekend_nights, stays_in_week_nights, adults, children, babies, meal, country, market_segment, distribution_channel, is_repeated_guest, previous_cancellations, previous_bookings_not_canceled, reserved_room_type, assigned_room_type, booking_changes, deposit_type, days_in_waiting_list, customer_type, adr, required_car_parking_spaces, total_of_special_requests, reservation_status, reservation_status_date]
Index: []

[0 rows x 31 columns]

There are no duplicates

# 3. Renaming Columns

In [13]:
# adr column doesn't have a clear name, so let's renaming to avg_daily_price that is price per room

In [14]:
df.rename(columns={'adr' : 'avg_daily_price'}, inplace = True)

In [15]:
df.head()

index         hotel  is_canceled  lead_time  arrival_date_year  \
0      0  Resort Hotel            0        342               2015   
1      1  Resort Hotel            0        737               2015   
2      2  Resort Hotel            0          7               2015   
3      3  Resort Hotel            0         13               2015   
4      4  Resort Hotel            0         14               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  assigned_room_type  \
0                        0                     0  ...                   C   
1                        0                     0  ...                   C   
2                        0                     1  ...                   C   
3                        0                     1  ...                   A   
4                        0                     2  ...                   A   

   booking_changes  deposit_type days_in_waiting_list customer_type  \
0                3    No Deposit                    0     Transient   
1                4    No Deposit                    0     Transient   
2                0    No Deposit                    0     Transient   
3                0    No Deposit                    0     Transient   
4                0    No Deposit                    0     Transient   

  avg_daily_price required_car_parking_spaces  total_of_special_requests  \
0             0.0                           0                          0   
1             0.0                           0                          0   
2            75.0                           0                          0   
3            75.0                           0                          0   
4            98.0                           0                          1   

   reservation_status  reservation_status_date  
0           Check-Out                 01-07-15  
1           Check-Out                 01-07-15  
2           Check-Out                 02-07-15  
3           Check-Out                 02-07-15  
4           Check-Out                 03-07-15  

[5 rows x 31 columns]

## 4. Data Understanding

In [16]:
# Basic descriptive statistical analysis
df.describe()

index    is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000  119390.000000      119390.000000   
mean    59694.500000       0.370416     104.011416        2016.156554   
std     34465.068657       0.482918     106.863097           0.707476   
min         0.000000       0.000000       0.000000        2015.000000   
25%     29847.250000       0.000000      18.000000        2016.000000   
50%     59694.500000       0.000000      69.000000        2016.000000   
75%     89541.750000       1.000000     160.000000        2017.000000   
max    119389.000000       1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes  days_in_waiting_list  avg_daily_price  \
count    119390.000000         119390.000000    119390.000000   
mean          0.221124              2.321149       101.831122   
std           0.652306             17.594721        50.535790   
min           0.000000              0.000000        -6.380000   
25%           0.000000              0.000000        69.290000   
50%           0.000000              0.000000        94.575000   
75%           0.000000              0.000000       126.000000   
max          21.000000            391.000000      5400.000000   

       required_car_parking_spaces  total_of_special_requests  
count                119390.000000              119390.000000  
mean                      0.062518                   0.571363  
std                       0.245291                   0.792798  
min                       0.000000                   0.000000  
25%                       0.000000                   0.000000  
50%                       0.000000                   0.000000  
75%            

After analysing the basic descriptive statistical analysis every numerical variable looks ok

In [17]:
df['arrival_date_year'].value_counts(dropna = False)

arrival_date_year
2016    56707
2017    40687
2015    21996
Name: count, dtype: int64

In [18]:
df['arrival_date_week_number'].value_counts(dropna = False)

arrival_date_week_number
33    3580
30    3087
32    3045
34    3040
18    2926
21    2854
28    2853
17    2805
20    2785
29    2763
42    2756
31    2741
41    2699
15    2689
27    2664
25    2663
38    2661
23    2621
35    2593
39    2581
22    2546
24    2498
13    2416
16    2405
19    2402
40    2397
26    2391
43    2352
44    2272
14    2264
37    2229
8     2216
36    2167
10    2149
9     2117
7     2109
12    2083
11    2070
45    1941
53    1816
49    1782
47    1685
46    1574
6     1508
50    1505
48    1504
4     1487
5     1387
3     1319
2     1218
52    1195
1     1047
51     933
Name: count, dtype: int64

In [19]:
df['arrival_date_day_of_month'].value_counts(dropna = False)

arrival_date_day_of_month
17    4406
5     4317
15    4196
25    4160
26    4147
9     4096
12    4087
16    4078
2     4055
19    4052
20    4032
18    4002
24    3993
28    3946
8     3921
3     3855
30    3853
6     3833
14    3819
27    3802
21    3767
4     3763
13    3745
7     3665
1     3626
23    3616
11    3599
22    3596
29    3580
10    3575
31    2208
Name: count, dtype: int64

In [20]:
df['adults'].value_counts(dropna = False)

adults
2     89680
1     23027
3      6202
0       403
4        62
26        5
27        2
20        2
5         2
40        1
50        1
55        1
6         1
10        1
Name: count, dtype: int64

0 has 403 records, it does not make sens, since children and babies do not travel/ spend holidays alone

In [21]:
df['children'].value_counts(dropna = False)

children
0.0     110796
1.0       4861
2.0       3652
3.0         76
NaN          4
10.0         1
Name: count, dtype: int64

In [22]:
df['babies'].value_counts(dropna = False)

babies
0     118473
1        900
2         15
10         1
9          1
Name: count, dtype: int64

In [23]:
df['is_repeated_guest'].value_counts(dropna = False)

is_repeated_guest
0    115580
1      3810
Name: count, dtype: int64

In [24]:
df['previous_cancellations'].value_counts(dropna = False)

previous_cancellations
0     112906
1       6051
2        116
3         65
24        48
11        35
4         31
26        26
25        25
6         22
19        19
5         19
14        14
13        12
21         1
Name: count, dtype: int64

In [25]:
df['booking_changes'].value_counts(dropna = False)

booking_changes
0     101314
1      12701
2       3805
3        927
4        376
5        118
6         63
7         31
8         17
9          8
10         6
13         5
14         5
15         3
16         2
17         2
12         2
11         2
20         1
21         1
18         1
Name: count, dtype: int64

In [26]:
df['days_in_waiting_list'].value_counts(dropna = False)

days_in_waiting_list
0      115692
39        227
58        164
44        141
31        127
        ...  
116         1
109         1
37          1
89          1
36          1
Name: count, Length: 128, dtype: int64

In [27]:
df['avg_daily_price'].value_counts(dropna = False)

avg_daily_price
62.00     3754
75.00     2715
90.00     2473
65.00     2418
0.00      1959
          ... 
89.43        1
63.07        1
55.69        1
49.51        1
157.71       1
Name: count, Length: 8879, dtype: int64

0 has 1959 records, it does not make sense, since the price per room should be greater than 0

In [28]:
df['required_car_parking_spaces'].value_counts(dropna = False)

required_car_parking_spaces
0    111974
1      7383
2        28
3         3
8         2
Name: count, dtype: int64

In [29]:
df['total_of_special_requests'].value_counts(dropna = False)

total_of_special_requests
0    70318
1    33226
2    12969
3     2497
4      340
5       40
Name: count, dtype: int64

In [30]:
# Exporting Data cleaned
df.to_csv(os.path.join(path, '02 Data','Prepared Data', 'hotel_bookings_cleaned.csv'))